code/modules/mining/gemstones.dm

```diff
@@ -0,0 +1,639 @@
+
+#define GEM_RUBY "ruby"
+#define GEM_TOPAZ "topaz"
+#define GEM_SAPPHIRE "sapphire"
+#define GEM_AMETHYST "amethyst"
+#define GEM_EMERALD "emerald"
+#define GEM_DIAMOND "diamond"
+#define GEM_MIXED "mixed"
+
+#define GEM_SMALL "small"
+#define GEM_MEDIUM "medium"
+#define GEM_LARGE "large"
+#define GEM_ENORMOUS "enormous"
+
+/obj/item/precious
+	name = "valuable object"
+	desc = "An object of value, made of crystallized coder tears."
+	var/maxstack
+	var/stack_contents = list()
+	var/pickup = 0
+	var/stacksize = 1
+	var/plural
+
+/obj/item/precious/gemstone
+	name = "gemstone"
+	desc = "An impossibly generic gemstone. How do you even have this?"
+	icon = 'icons/obj/gemstones.dmi'
+	var/gemcolor
+	var/gemtype
+	var/gemsize
+
+/obj/item/precious/gemstone/Initialize(mapload, amount)
+	. = ..()
+	if(amount)
+		stacksize = amount
+		update_gem()
+
+/obj/item/precious/gemstone/mixed/Initialize(mapload, amount, kind)
+	ASSERT(kind)
+	stack_contents[kind] = amount
+	. = ..(mapload, amount)
+
+/obj/item/precious/gemstone/attackby(obj/item/precious/gemstone/G, mob/user)
+	if (!istype(G))
+		return ..()
+	if(G.gemtype == GEM_MIXED)
+		to_chat(user, span("notice", "Try doing it the other way - put an unmixed gem into the pile, instead of the pile onto an organized gem stack."))
+		return
+	if(src.gemsize != G.gemsize)
+		to_chat(user, span("notice", "You can't mix gem sizes. It would be too disorganized!"))
+		return
+	if(src.gemsize == GEM_ENORMOUS)
+		to_chat(user, span("notice", "That gem is far too big to stack!"))
+		return
+	if(src.gemtype == G.gemtype)
+		gemstacker(G, user)
+	else
+		gemmixer(G, user)
+
+/obj/item/precious/gemstone/AltClick(mob/user)
+	if(src.gemsize == GEM_ENORMOUS)
+		user.put_in_hands(src)
+		return
+	if(src.stacksize == 1) // just pick it up...
+		user.put_in_hands(src)
+		return
+	if(src.gemtype == GEM_MIXED)
+		to_chat(user, span("notice", "Try disassembling the whole stack instead, if you want to separate it."))
+		return
+	else 
+		to_chat(user, span("notice", "You take a [src.gemsize] [src.gemtype] from the pile."))
```

Comment:
"You take **a enormous emerald** from the pile."

Reply:
it should display as "a small green" gemstone.